src/ibek/definition.py

```diff
@@ -0,0 +1,168 @@
+"""
+The Definition class describes what a given support module can instantiate.
+"""
+
+from __future__ import annotations
+
+from enum import Enum
+from typing import Any, Mapping, Optional, Sequence, Union
+
+from pydantic import Field, PydanticUndefinedAnnotation
+from typing_extensions import Literal
+
+from .args import Arg, IdArg, Value
+from .globals import BaseSettings
+from .sub_entity import SubEntity
+
+
+def default(T: type):
+    """
+    defines a default type which may be
+    """
+    return Field(
+        Union[Optional[T], PydanticUndefinedAnnotation],
+        description="If given, instance doesn't supply argument, what value should be used",
+    )
+
+
+class When(Enum):
+    first = "first"
+    every = "every"
+    last = "last"
+
+
+class Database(BaseSettings):
+    """
+    A database file that should be loaded by the startup script and its args
+    """
+
+    file: str = Field(
+        description="Filename of the database template in <module_root>/db"
+    )
+
+    enabled: str = Field(
+        description="Set to False to disable loading this database", default="True"
+    )
+
+    args: Mapping[str, Optional[str]] = Field(
+        description=(
+            "Dictionary of args and values to pass through to database. "
+            "A value of None is equivalent to ARG: '{{ ARG }}'. "
+            "See `UTILS.render_map` for more details."
+        )
+    )
+
+
+class EnvironmentVariable(BaseSettings):
+    """
+    An environment variable that should be set in the startup script
+    """
+
+    name: str = Field(description="Name of environment variable")
+    value: str = Field(description="Value to set")
+
+
+class Comment(BaseSettings):
+    """
+    A script snippet that will have '# ' prepended to every line
+    for insertion into the startup script
+    """
+
+    type: Literal["comment"] = "comment"
+    when: When = Field(description="One of first / every / last", default="every")
+    value: str = Field(
+        description="A comment to add into the startup script", default=""
+    )
+
+
+class Text(BaseSettings):
+    """
+    A script snippet to insert into the startup script
+    """
+
+    type: Literal["text"] = "text"
+    when: str = Field(description="One of first / every / last", default="every")
+    value: str = Field(description="raw text to add to the startup script", default="")
+
+
+Script = Sequence[Union[Text, Comment]]
+
+
+class EntityPVI(BaseSettings):
```

Comment:
Could this be moved next to `Entity`?

Reply:
I think that is wrong. Entity is a base class whose concrete subclasses are dynamically built from a Definition instance. EntityPVI is a concrete class that is part of Defintion and contributes to what an Entity class looks like (phew!)

Reply:
Hmm. Yes you are right. Maybe I shouldn't have called it `EntityPVI` then?

Reply:
I'm OK with the name - it is the pvi for the enities that will get generated by this definition. But also just calling it PVI like its field name would also be fine.

Reply:
I'm changing this to PVI - more in keeping with the other classes in a definition.

Reply:
This changes the schema metadata so I'm also taking the opportunity to change Definition to EntityDefinition.